### PR TITLE
[Torch] refactor Engine and Backends

### PR DESCRIPTION
**Pull request by fortianyou**
Related to ONNX backends opensource: https://github.com/alibaba/BladeDISC/issues/157

The PR features EngineClass(A TorchScript CustomClassHolder) and EngineInterface of TorchBlade's backend, which is responsible to do PyTorch backends engine building and serialization.

After that, we will subclass the DISC engine and other accelerators' engines against the interface of TorchBlade's backend.

This PR also refactors the implementation of the BladeDISC backend.

### CODE REVIEW

**Comment by Orion34-lanbo** [comment]
What is the purpose of this refactor?

**Comment by Yancey1989** [comment]
Better to add an issue to descript this refactor?

File: pytorch_blade/src/compiler/backends/engine_interface.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+namespace {
+struct EngineCreatorRegistry {
+  std::unordered_map<std::string, EngineCreatorRegister::EngineCreator> lut;
+
+  EngineCreatorRegistry& RegisterBackend(
+      const std::string& name,
+      const EngineCreatorRegister::EngineCreator& creator) {
+    auto found = lut.find(name);
+    if (found != lut.end()) {
+      LOG(WARNING) << "Overriding already registered EngineCreator for backend "
```
Comment (Yancey1989) [inline pytorch_blade/src/compiler/backends/engine_interface.cpp:30]: what if registry multiple backends? If that causes unpredictable behavior, I prefer using `LOG(FATAL)`.

File: pytorch_blade/src/compiler/backends/engine_interface.h [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  // To avoid changing the fields, please
+  // add new debug fields to the debug attr dict
+  using State = EngineState;
+  virtual const State& GetState() const = 0;
+
+  static const char* GetBackendName() {
+    return "Engine";
+  }
+
+  virtual torch::List<torch::Tensor> Forward(
```
Comment (Yancey1989) [inline pytorch_blade/src/compiler/backends/engine_interface.h:57]: Do we want to add a `Backward` function?

File: pytorch_blade/src/compiler/backends/engine_interface.cpp (hunk 1 above)
Comment (fortianyou) [reply pytorch_blade/src/compiler/backends/engine_interface.cpp:30]: An elder registered one would be eliminated. I think It's good to use `LOG(FATAL)`.

File: pytorch_blade/src/compiler/backends/engine_interface.h (hunk 2 above)
Comment (fortianyou) [reply pytorch_blade/src/compiler/backends/engine_interface.h:57]: Is the `Backward` here meant for training? I don't think we need it in EngineInterface, because the EngineInterface just runs what it's given. Something like gradients and parameters is transparent to EngineInterface.

However, the naming `Forward` here might be ambiguous, I  will rename it to `Execute`.

Comment (Yancey1989) [reply pytorch_blade/src/compiler/backends/engine_interface.h]: Agree with the name `Execute` :)

File: pytorch_blade/src/compiler/backends/backend_input_outputs.cpp [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+
+#include "backend_input_outputs.h"
+#include "compiler/jit/shape_type_spec.h"
+#include "compiler/jit/tool_funcs.h"
+
+namespace torch {
+namespace blade {
+namespace backends {
+
+bool DynamicRanges::Validate(int inp_nums) {
```
Comment (Yancey1989) [inline pytorch_blade/src/compiler/backends/backend_input_outputs.cpp:20]: It's better to keep a pull request small and clean. it seems that `DynamicRanges`  is not necessary for this refactor task.

ref: https://developers.google.com/blockly/guides/modify/contribute/write_a_good_pr

**Review by Yancey1989 — APPROVED** [review]
LGTM
